src/future/shared.rs

```diff
@@ -105,188 +122,165 @@ impl<F> Future for Shared<F>
+
+                return Ok(Async::NotReady);
+            }
+            COMPLETE => {
+                return unsafe { self.clone_result().map(Async::Ready) };
+            }
+            POISONED => panic!("inner future panicked during poll"),
+            _ => unreachable!(),
+        }
+
+        loop {
+            struct Reset<'a>(&'a AtomicUsize);
+
+            impl<'a> Drop for Reset<'a> {
+                fn drop(&mut self) {
+                    use std::thread;
+
+                    if thread::panicking() {
+                        self.0.store(POISONED, SeqCst);
+                    }
                 }
-                return Ok(Async::NotReady)
             }
-            State::Done(Ok(ref v)) => return Ok(SharedItem { item: v.clone() }.into()),
-            State::Done(Err(ref e)) => return Err(SharedError { error: e.clone() }.into()),
-        };
 
-        let new_state = State::Polling(unparker, Vec::new());
-        let (unparker, mut original_future) = match mem::replace(&mut *state, new_state) {
-            State::Waiting(unparker, original_future) => (unparker, original_future),
-            _ => unreachable!(),
-        };
-        drop(state);
-
-        let event = task::UnparkEvent::new(unparker.clone(), 0);
-        let new_state = match task::with_unpark_event(event, || original_future.poll()) {
-            Ok(Async::NotReady) => State::Waiting(unparker, original_future),
-            Ok(Async::Ready(v)) => State::Done(Ok(Arc::new(v))),
-            Err(e) => State::Done(Err(Arc::new(e))),
-        };
-
-        let (call_unpark, result) = match new_state {
-            State::Waiting(..) => (false, Ok(Async::NotReady)),
-            State::Polling(..) => unreachable!(),
-            State::Done(Ok(ref v)) => (true, Ok(SharedItem { item: v.clone() }.into())),
-            State::Done(Err(ref e)) => (true, Err(SharedError { error: e.clone() }.into())),
-        };
-
-        let mut state = self.inner.state.lock().unwrap();
-        match mem::replace(&mut *state, new_state) {
-            State::Polling(unparker, waiters) => {
-                if call_unpark { unparker.unpark() }
-                for waiter in waiters {
-                    waiter.unpark();
+            let _reset = Reset(&self.inner.unparker.state);
+
+            // Get a handle to the unparker
+            let unpark: Arc<Unpark> = self.inner.unparker.clone();
+
+            // Poll the future
+            match unsafe { (*self.inner.future.get()).poll_future(unpark) } {
+                Ok(Async::NotReady) => {
+                    // Not ready, try to release the handle
+                    match self.inner.unparker.state.compare_and_swap(POLLING, IDLE, SeqCst) {
+                        POLLING => {
+                            // Success
+                            return Ok(Async::NotReady);
+                        }
+                        REPOLL => {
+                            // Gotta poll again!
```

Comment:
Interesting! It seems plausible to me that this "loop on REPOLL" business fixes the bug that @stuhood has been encountering in the current implementation.